### PR TITLE
Implementing Flask Webserver

### PR DESCRIPTION
**Pull request by tanner-g**
WIP:

- files are currently being built in /config/webserver
- setup.sh is used to run locally with debug features on
- will be moved to a centos container for deployment

### CODE REVIEW

**Comment by tanner-g** [comment]
when static html/js/css etc is ready for testing, place them into the templates/ folder

**Comment by tanner-g** [comment]
Currently working on db integration. Flask now serves webpages and static content.

**Comment by tanner-g** [comment]
passwords are currently in plaintext in db... need to fix that

**Comment by tanner-g** [comment]
Update:
- can create users 
- can login with users

**Comment by tanner-g** [comment]
Still working on catching bad logins... and bad registration..

**Comment by tanner-g** [comment]
if we used encoding for usernames, and hashing for passwords, we could protect the db from sql injection. encode usernames to base64 and hash passwords for protection

**Comment by tanner-g** [comment]
Tests to Write:
- add user
- check user / password

**Comment by tanner-g** [comment]
TO DO : 
- add webserver Dockerfile to docker-compose

**Comment by JasonPuglisi** [comment]
Everything is working through `docker-compose up` now and current tests are passing. I'm using this branch to also refactor a few non-branch-related files.

**Comment by tanner-g** [comment]
Error checking on username input and password input is successful. Testing will still need to occur. Now enforces username and password lengths, and alphanumeric on username.

**Comment by tanner-g** [comment]
user page protection is coming along... its close.

**Comment by tanner-g** [comment]
using flask-login to handle session information.

File: config/webserver/emmental_flask/db.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  for result in cursor:
+    print(result)
+
+# insert/add user into mariadb
+def add_user(username, password):
+  dbconnection = get_db()
+  userID = generate_userID()
+  cursor = dbconnection.cursor()
+  dbusername = username
+  cursor.execute("INSERT INTO Users (userID, username, password) VALUES (%s,%s,%s)", (userID, dbusername, password))
```
Comment (JasonPuglisi) [inline config/webserver/emmental_flask/db.py]: Change the %s for the userID to UUID_TO_BIN(%s)

File: config/webserver/emmental_flask/db.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  if not password_check:
+    return False
+  else:
+    return True
+
+# generate and return 16 char userID
+def generate_userID():
+  dbconnection = get_db()
+  cursor = dbconnection.cursor()
+  query = 'SELECT LEFT(UUID(), 16);'
```
Comment (JasonPuglisi) [inline config/webserver/emmental_flask/db.py]: Don't select left, just `SELECT UUID();`

**Comment by nixs123** [comment]
Homepage.html


signup.html


login.html


index.html

File: config/webserver/emmental_flask/db.py (hunk 1 above)
Comment (tanner-g) [reply config/webserver/emmental_flask/db.py]: Converting to binary makes the fields larger than we really need to store. It's easier to just store the full 36 char UUID instead.

File: config/webserver/emmental_flask/db.py (hunk 2 above)
Comment (tanner-g) [reply config/webserver/emmental_flask/db.py]: Changed!

**Comment by tanner-g** [comment]
> Also utilize bcrypt for passwords
> 
> Add some tests for things

Next step is tests. Bcrypt only supports password lengths up to 72 chars. There is a security bug that was reported from "wrapping around". Since we have chosen this method for hashing, I am going to change the password requirements to a max 72 char password. It will work if you supply more characters, but Bcrypt truncates it. It is better than SHA because the salt is random for every password hashed. If it works for our purposes, it will be protected against rainbow attacks.

**Comment by tanner-g** [comment]
Functionality Completed: 

[x] Bcrypt for passwords (length 72 chars max)
[x] UUID (char 36) for userIDs

Up Next:
[ ] Unit Tests for password functions
[ ] Unit Tests for creating a user
[ ] Validation of flask-login, then unit test for this.

**Comment by JasonPuglisi** [comment]
Closes #23 
Closes #26 
Closes #27 
Closes #28

**Pull request merged**